gridded_etl_tools/utils/publish.py

```diff
@@ -793,35 +793,147 @@ def check_written_value(
+        Find the time in a selected raw dataset and filter down a list of local files to include only
+        files that contain that time
+
+        Parameters
+        ----------
+        original_files: tuple[str]
+            A list of raw files
+
+        raw_ds : xr.Dataset
+            The raw dataset used for filtering
+
+        Returns
+        -------
+        time_filtered_original_files : tuple[str]
+            A list of raw files, filtered to only contain files that contain the specified time
+        """
+        time_string = self.numpydate_to_py(np.atleast_1d(raw_ds[self.time_dim])[0]).date().isoformat()
+        time_filtered_original_files = [fil for fil in original_files if time_string in str(fil)]
+        return time_filtered_original_files
+
+    def filter_files_by_step(self, original_files: tuple[str], raw_ds: xr.Dataset) -> tuple[str]:
+        """
+        Find the step number in a selected raw dataset and filter down a list of local files to include only
+        files that contain that step number
+
+        Parameters
+        ----------
+        original_files: tuple[str]
+            A list of raw files, similarly filtered to only include files covering a specific forecast step.
+
+        raw_ds : xr.Dataset
+            The raw dataset used for filtering
+
+        Returns
+        -------
+        step_filtered_original_files : tuple[str]
+            A list of raw files, filtered to only contain files that contain the specified step number
+        """
+        step_string = "step-" + str(int(raw_ds["step"].values[0] / 3600000000000)) + "_"
+        step_filtered_original_files = [fil for fil in original_files if step_string in str(fil)]
+        return step_filtered_original_files
+
+    def filter_files_by_ensemble(
+        self, original_files: tuple[str], raw_ds: xr.Dataset
+    ) -> tuple[str]:  # pragma NO COVER
+        """
+        Find the ensemble number in a selected raw dataset and filter down a list of local files to include only
+        files that contain that ensemble
+
+        Parameters
+        ----------
+        original_files: tuple[str]
+            A list of raw files, similarly filtered to only include files covering a specific ensemble number.
+
+        raw_ds : xr.Dataset
+            The raw dataset used for filtering
+
+        Returns
+        -------
+        ensemble_filtered_original_files : tuple[str]
+            A list of raw files, filtered to only contain files that contain the specified ensemble
+        """
+        ensemble_string = "ensemble-" + str(raw_ds["ensemble"].values[0]) + "_"
+        ensemble_filtered_original_files = [fil for fil in original_files if ensemble_string in str(fil)]
+        return ensemble_filtered_original_files
+
     def binary_search_for_file(
-        self, target_datetime: np.datetime64, time_dim: str, possible_files: list[str] | None = None
+        self, target_datetime: np.datetime64, time_dim: str, possible_files: list[str]
```

Comment:
I realize I'm late to the party with this comment, but I'm not sure a binary search is the best solution here. Wouldn't it be better to just choose the files beforehand at the same time the coordinates are chosen? Then we could just iterate over the list of coordinates+files. If the pure randomness is a problem, we could use a different version of randomization where we go in order over the sorted list of files at the start of the post parse check, randomly choosing ones to check, and randomly choosing a coordinate or coordinates in the files.